### PR TITLE
Add timelock on lock and unlock functions (Issue #105)

### PR DESCRIPTION
**Pull request by alebaffa**

### CODE REVIEW

File: contracts/FragToken.sol [hunk 1]
```diff
@@ -9,15 +9,24 @@ import "openzeppelin-solidity/contracts/access/Ownable.sol";
 
 contract FRAGToken is ERC20, ERC20Permit, Ownable {
     uint8 constant DECIMALS = 12; // Preferred for Fragnova (Substrate)
-    uint256 constant INITIAL_SUPPLY = 10_000_000_000 * (10**DECIMALS);
+    uint256 constant INITIAL_SUPPLY = 10_000_000_000 * (10**DECIMALS); 
+    uint256 private _lockCooldown = 1 weeks; // Roughly 1 week
```
Comment (sinkingsugar) [inline contracts/FragToken.sol:13]: can remove the comment 😄

File: contracts/FragToken.sol [hunk 2]
```diff
@@ -64,20 +73,40 @@ contract FRAGToken is ERC20, ERC20Permit, Ownable {
...
+            _lockTime[msg.sender] = block.timestamp + (4 * _TIMELOCK);
+
+        if(period == Period.ThreeMonths)
+            _lockTime[msg.sender] = block.timestamp + (12 * _TIMELOCK);
+
+        if(period == Period.SixMonths)
+            _lockTime[msg.sender] = block.timestamp + (24 * _TIMELOCK);
+
+        if(period == Period.OneYear)
+            _lockTime[msg.sender] = block.timestamp + (52 * _TIMELOCK);
```
Comment (sinkingsugar) [inline contracts/FragToken.sol:91]: I think we need an if/elseif construct with a final else as I'm not 100% sure if solidity will handle unknown cases.
Transactions can be crafted with any bytes layout...

File: contracts/FragToken.sol [hunk 3]
```diff
@@ -64,20 +73,40 @@ contract FRAGToken is ERC20, ERC20Permit, Ownable {
...
         // emit total amount of locked tokens
-        emit Lock(msg.sender, signature, _locksAmount[msg.sender]);
+        emit Lock(msg.sender, signature, _locksAmount[msg.sender], _lockTime[msg.sender]);
     }
 
     function unlock(bytes calldata signature) external {
         require(
             block.number > _locksBlock[msg.sender] + _lockCooldown,
-            "Lock didn't expire"
+            "Lock cooldown didn't expire"
```
Comment (sinkingsugar) [inline contracts/FragToken.sol:104]: actually cooldown should be removed no? we just need to use lock now

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

File: contracts/FragToken.sol (hunk 3 above)
Comment (alebaffa) [reply contracts/FragToken.sol:104]: oh well, yes of course!

File: contracts/FragToken.sol (hunk 1 above)
Comment (alebaffa) [reply contracts/FragToken.sol]: Done, I removed completely the cooldown variable and setter function.

File: contracts/FragToken.sol (hunk 2 above)
Comment (alebaffa) [reply contracts/FragToken.sol]: Yes, correct.

File: contracts/FragToken.sol [hunk 4]
```diff
@@ -64,20 +68,36 @@ contract FRAGToken is ERC20, ERC20Permit, Ownable {
...
+            _lockTime[msg.sender] = block.timestamp + (4 * _TIMELOCK);
+
+        else if(period == Period.ThreeMonths)
+            _lockTime[msg.sender] = block.timestamp + (12 * _TIMELOCK);
+
+        else if(period == Period.SixMonths)
+            _lockTime[msg.sender] = block.timestamp + (24 * _TIMELOCK);
+
+        else {
+            if(period == Period.OneYear)
```
Comment (sinkingsugar) [inline contracts/FragToken.sol:85]: move this up and `else` should simply fail?

Comment (alebaffa) [reply contracts/FragToken.sol:92]: I introduced an additional `require` instead to check the period in input. It is where it fails. I think it's cleaner and also more coherent with how we already handle the inputs.

File: contracts/FragToken.sol [hunk 5]
```diff
@@ -49,6 +49,13 @@ contract FRAGToken is ERC20, ERC20Permit, Ownable {
 
     function lock(uint256 amount, bytes calldata signature, Period period) external {
         require(amount > 0, "Amount must be greater than 0");
+        require(
```
Comment (sinkingsugar) [inline contracts/FragToken.sol:52]: gas wise this is probably bad as you then repeat the same logic under

File: contracts/FragToken.sol (hunk 4 above)
Comment (sinkingsugar) [reply contracts/FragToken.sol:92]: there is no clean code with ethereum contract, everything costs gas. see my above comment 😭😄

Comment (alebaffa) [reply contracts/FragToken.sol:92]: I didn't consider gas optimization. My very bad. Let me try few different approaches. Thank you for your comment!

File: contracts/FragToken.sol (hunk 5 above)
Comment (alebaffa) [reply contracts/FragToken.sol:52]: Right, they could just be in one single revert.

File: contracts/FragToken.sol [hunk 6]
```diff
@@ -9,15 +9,23 @@ import "openzeppelin-solidity/contracts/access/Ownable.sol";
...
+        TwoWeeks,
+        OneMonth,
+        ThreeMonths,
+        SixMonths,
+        OneYear
+    }
 
     // Fragnova chain will listen to those events
-    event Lock(address indexed sender, bytes signature, uint256 amount);
+    event Lock(address indexed sender, bytes signature, uint256 amount, uint256 timelock);
```
Comment (sinkingsugar) [inline contracts/FragToken.sol:28]: Now that I see this, we need to make sure that clamor side this `timelock` is ignored when `Unlock` happens, not sure if that's the case. Also event HASH has changed (I think it's a constant clamor side)

**Review by sinkingsugar — APPROVED** [review]
Ok but do we need this forge-std module? I don't see it used.

**Comment by alebaffa** [comment]
> Ok but do we need this forge-std module? I don't see it used.

No, it shouldn't be there! Thanks! Just removed it.

**Comment by sinkingsugar** [comment]
I see one conflict, not sure why

File: contracts/FragToken.sol (hunk 6 above)
Comment (alebaffa) [reply contracts/FragToken.sol:28]: Correct, the event hash has changed too. Need to change on Clamor side.

**Pull request merged**